ops/testing.py

```diff
@@ -1075,6 +1076,75 @@ def __init__(self, resource_name):
         self.name = resource_name
 
 
+@_record_calls
+class _TestingConfig(dict):
+    """Represents the Juju Config."""
+    _supported_types = {
+        'string': str,
+        'boolean': bool,
+        'int': int,
+        'float': float
+    }
+
+    def __init__(self, config: dict):
+        super().__init__()
+        self._spec = config
+        self._defaults = self._load_defaults(config)
+
+        for key, value in self._defaults.items():
+            if value is None:
+                continue
+            self._config_set(key, value)
+
+    @staticmethod
+    def _load_defaults(charm_config: dict):
+        """Load default values from config.yaml.
+
+        Handle the case where a user doesn't supply explicit config snippets.
+        """
+        if not charm_config:
+            return {}
+        charm_config = charm_config.get('options', {})
+        return {key: value.get('default', None) for key, value in charm_config.items()}
+
+    def _config_set(self, key, value):
+        # this is only called by the harness itself
+        # we don't do real serialization/deserialization, but we do check that the value
+        # has the expected type.
+        if type(value) not in self._supported_types.values():
+            raise RuntimeError('cannot set {} to {}; type {} is not '
+                               'supported by juju'.format(key, value, type(value)))
```

Comment:
This check seems redundant with the following checks which in combination should catch the same error case.

Reply:
true that! good catch :)